va/validation-authority.go

```diff
@@ -637,33 +641,75 @@ func (va *ValidationAuthorityImpl) checkCAARecords(ctx context.Context, identifi
 	if err != nil {
 		return
 	}
+
 	if caaSet == nil {
 		// No CAA records found, can issue
+		va.stats.Inc("VA.CAA.None", 1, 1.0)
 		present = false
 		valid = true
 		return
-	} else if caaSet.criticalUnknown() {
-		present = true
+	}
+
+	present = true
+
+	// Record stats on directives not currently processed.
+	if len(caaSet.Iodef) > 0 {
+		va.stats.Inc("VA.CAA.WithIodef", 1, 1.0)
+	}
+
+	if caaSet.criticalUnknown() {
+		// Contains unknown critical directives.
+		va.stats.Inc("VA.CAA.UnknownCritical", 1, 1.0)
 		valid = false
 		return
-	} else if len(caaSet.Issue) > 0 || len(caaSet.Issuewild) > 0 {
-		present = true
-		var checkSet []*dns.CAA
-		if strings.SplitN(hostname, ".", 2)[0] == "*" {
-			checkSet = caaSet.Issuewild
-		} else {
-			checkSet = caaSet.Issue
-		}
-		for _, caa := range checkSet {
-			if caa.Value == va.IssuerDomain {
-				valid = true
-				return
-			}
-		}
+	}
 
-		valid = false
+	if len(caaSet.Unknown) > 0 {
+		va.stats.Inc("VA.CAA.WithUnknownNoncritical", 1, 1.0)
+	}
+
+	if len(caaSet.Issue) == 0 {
+		// Although CAA records exist, none of them pertain to issuance in this case.
+		// (e.g. there is only an issuewild directive, but we are checking for a
+		// non-wildcard identifier, or there is only an iodef or non-critical unknown
+		// directive.)
+		va.stats.Inc("VA.CAA.NoneRelevant", 1, 1.0)
+		valid = true
 		return
 	}
 
+	// There are CAA records pertaining to issuance in our case. Note that this
+	// includes the case of the unsatisfiable CAA record value ";", used to
+	// prevent issuance by any CA under any circumstance.
+	//
+	// Our CAA identity must be found in the chosen checkSet.
+	for _, caa := range caaSet.Issue {
+		if extractIssuerDomain(caa) == va.IssuerDomain {
+			va.stats.Inc("VA.CAA.Authorized", 1, 1.0)
+			valid = true
+			return
+		}
+	}
+
+	// The list of authorized issuers is non-empty, but we are not in it. Fail.
+	va.stats.Inc("VA.CAA.Unauthorized", 1, 1.0)
+	valid = false
 	return
 }
+
+// Given a CAA record, assume that the Value is in the issue/issuewild format,
+// that is, a domain name with zero or more additional key-value parameters.
+// Returns the domain name, which may be "" (unsatisfiable).
+func extractIssuerDomain(caa *dns.CAA) string {
+	v := caa.Value
+	v = strings.Trim(v, " \t") // Value can start and end with whitespace.
```

Comment:
What is considered whitespace is determined precisely by the ABNF in the RFC. TrimSpace removes arbitrary Unicode space characters.